File: 238-product-of-array-except-self/product-of-array-except-self.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        // create prefix and suffix products 
        // loop through nums and multiply prefix[i-1] and suffix[i+1] values for current product
        // if i - 1 or i + 1 out of bounds, multiply by 1
        
        val prefix = MutableList<Int>(nums.size) {0}
        val suffix = MutableList<Int>(nums.size) {0}
        val products = IntArray(nums.size) {0}
        var product = 1


        nums.forEachIndexed { idx, num -> 
            product *= num
            prefix[idx] = product
        }

        product = 1
        nums.toMutableList().asReversed().forEachIndexed { idx, num -> 
            product *= num
            print("$idx ")
            suffix[nums.size - 1 - idx] = product
        }

        println(prefix)
        println(suffix)

        nums.forEachIndexed { idx, value -> 
            if (idx == 0) {
                product = 1 * suffix[idx + 1]
            } else if (idx == nums.size - 1) {
                product = 1 * prefix[idx - 1]
            } else {
                product = suffix[idx + 1] * prefix[idx - 1]
            }
            products[idx] = product
        }
        println(products)
        return products
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix  = [1] * len(nums)
        postfix = [1] * len(nums)
        output  = [1] * len(nums)

        # calculate prefixes
        preProduct = 1
        postProduct = 1
        for i in range(len(nums)):
            preProduct *= nums[i]
            prefix[i] = preProduct
            
            postProduct *= nums[len(nums) - i - 1]            
            postfix[len(nums) - i - 1] = postProduct
        print(postfix)

        for i in range(len(nums)):
            left  = prefix[i-1] if i - 1 >= 0 else 1
            right = postfix[i+1] if i + 1 < len(nums) else 1
            output[i] = left * right

        return output
"""
create prefix and postfix product array 

loop through nums and basically just take the index - 1 * index + 1 and that's the current product except self

i.e. nums = [1,2,3,4]
prefix    = [1,2,6,24]
postfix   = [24,24,12,4]

in the event i-1 or i+1 is OoB, than assume it multiplies by 1
"""